src/cryptography/hazmat/primitives/serialization/ssh.py

```diff
@@ -703,6 +718,256 @@ def _serialize_ssh_private_key(
+        _signature: memoryview,
+        _tbs_cert_body: memoryview,
+        _cert_key_type: bytes,
+        _cert_body: memoryview,
+    ):
+        self._nonce = _nonce
+        self._public_key = _public_key
+        self._serial = _serial
+        try:
+            self._type = SSHCertificateType(_cctype)
+        except ValueError:
+            raise ValueError("Invalid certificate type")
+        self._key_id = _key_id
+        self._valid_principals = _valid_principals
+        self._valid_after = datetime.datetime.utcfromtimestamp(_valid_after)
+        self._valid_before = datetime.datetime.utcfromtimestamp(_valid_before)
+        self._critical_options = _critical_options
+        self._extensions = _extensions
+        self._sig_type = _sig_type
+        self._sig_key = _sig_key
+        self._signature = _signature
+        self._cert_key_type = _cert_key_type
+        self._cert_body = _cert_body
+        self._tbs_cert_body = _tbs_cert_body
+
+    @property
+    def nonce(self) -> bytes:
+        return bytes(self._nonce)
+
+    def public_key(self) -> _SSH_CERT_PUBLIC_KEY_TYPES:
+        # make mypy happy until we remove DSA support entirely and
+        # the underlying union won't have a disallowed type
+        assert not isinstance(self._public_key, dsa.DSAPublicKey)
+        return self._public_key
+
+    @property
+    def serial(self) -> int:
+        return self._serial
+
+    @property
+    def type(self) -> SSHCertificateType:
+        return self._type
+
+    @property
+    def key_id(self) -> bytes:
+        return bytes(self._key_id)
+
+    @utils.cached_property
+    def valid_principals(self) -> typing.List[bytes]:
+        rest = self._valid_principals
+        principals = []
+        while len(rest) > 0:
+            principal, rest = _get_sshstr(rest)
+            principals.append(bytes(principal))
+
+        return principals
+
+    @property
+    def valid_before(self) -> datetime.datetime:
+        return self._valid_before
+
+    @property
+    def valid_after(self) -> datetime.datetime:
+        return self._valid_after
+
+    @utils.cached_property
+    def critical_options(self) -> typing.Dict[bytes, bytes]:
+        return self._parse_exts_opts(self._critical_options)
+
+    @utils.cached_property
+    def extensions(self) -> typing.Dict[bytes, bytes]:
+        return self._parse_exts_opts(self._extensions)
+
+    def _parse_exts_opts(
+        self, exts_opts: memoryview
+    ) -> typing.Dict[bytes, bytes]:
+        result: typing.Dict[bytes, bytes] = {}
+        while len(exts_opts) > 0:
+            name, exts_opts = _get_sshstr(exts_opts)
+            bname: bytes = bytes(name)
+            if bname in result.keys():
+                raise ValueError("Duplicate extension name")
+            # dicts in pypy & python 3.6+ are insertion ordered so hooray
+            if len(result) > 0 and bname < list(result.keys())[-1]:
+                raise ValueError("Extensions not lexically sorted")
```

Comment:
This is way too complicatd. Just store the previous value in a local key.

The way you've written this makes the algorithm O(n^2).